File: python_bot/cogs/role_reactions.py
import discord
from discord.ext import commands
from discord import app_commands
from typing import Optional

class RoleReactionsCog(commands.Cog):
    """Role reaction system for self-assignable roles"""
    
    def __init__(self, bot):
        self.bot = bot
    
    @app_commands.command(name="reactionrole", description="Set up a reaction role on a message")
    @app_commands.describe(
        channel="The channel where the message is located",
        message_id="The ID of the message to add the reaction role to",
        emoji="The emoji to react with",
        role="The role to give when users react"
    )
    @app_commands.default_permissions(manage_roles=True)
    async def setup_reaction_role(self, interaction: discord.Interaction, 
                                  channel: discord.TextChannel, message_id: str, 
                                  emoji: str, role: discord.Role):
        """Set up a reaction role"""
        
        # Check bot permissions
        if not interaction.guild.me.guild_permissions.manage_roles:
            await interaction.response.send_message("❌ I need the Manage Roles permission to set up reaction roles!", ephemeral=True)
            return
        
        # Check if bot's role is high enough
        if role >= interaction.guild.me.top_role:
            await interaction.response.send_message(f"❌ I cannot manage the {role.mention} role because it's higher than or equal to my highest role!", ephemeral=True)
            return
        
        # Check if user's role is high enough
        if role >= interaction.user.top_role and not interaction.user.guild_permissions.administrator:
            await interaction.response.send_message(f"❌ You cannot set up reaction roles for {role.mention} because it's higher than or equal to your highest role!", ephemeral=True)
            return
        
        try:
            # Try to fetch the message from the specified channel
            message = await channel.fetch_message(int(message_id))
        except discord.NotFound:
            await interaction.response.send_message(f"❌ Message not found in {channel.mention}! Make sure the message ID is correct.", ephemeral=True)
            return
        except discord.Forbidden:
            await interaction.response.send_message(f"❌ I don't have permission to access messages in {channel.mention}!", ephemeral=True)
            return
        except ValueError:
            await interaction.response.send_message("❌ Invalid message ID format!", ephemeral=True)
            return
        
        if not self.bot.db_pool:
            await interaction.response.send_message("❌ Database connection not available!", ephemeral=True)
            return
        
        try:
            async with self.bot.db_pool.acquire() as conn:
                # Check if this emoji/message combo already exists
                existing = await conn.fetchrow("""
                    SELECT id FROM role_reactions 
                    WHERE message_id = $1 AND emoji = $2 AND is_active = true
                """, message_id, emoji)
                
                if existing:
                    await interaction.response.send_message(f"❌ A reaction role with {emoji} already exists on that message!", ephemeral=True)
                    return
                
                # Add to database
                await conn.execute("""
                    INSERT INTO role_reactions 
                    (server_id, channel_id, message_id, emoji, role_id, is_active)
                    VALUES ($1, $2, $3, $4, $5, $6)
                """, str(interaction.guild.id), str(channel.id), 
                     message_id, emoji, str(role.id), True)
                
                # Add reaction to the message
                try:
                    await message.add_reaction(emoji)
                except discord.HTTPException:
                    await interaction.response.send_message(f"❌ Failed to add reaction! Make sure the emoji is valid.", ephemeral=True)
                    # Remove from database since we couldn't add the reaction
                    await conn.execute("""
                        DELETE FROM role_reactions 
                        WHERE message_id = $1 AND emoji = $2
                    """, message_id, emoji)
                    return
                
                embed = discord.Embed(
                    title="✅ Reaction Role Created",
                    description=f"React with {emoji} on [this message]({message.jump_url}) to get {role.mention}!",
                    color=discord.Color.green()
                )
                embed.add_field(name="Message ID", value=message_id, inline=True)
                embed.add_field(name="Emoji", value=emoji, inline=True)
                embed.add_field(name="Role", value=role.mention, inline=True)
                
                await interaction.response.send_message(embed=embed)
                print(f"Created reaction role: {emoji} -> {role.name} on message {message_id}")
                
        except Exception as e:
            print(f"Error creating reaction role: {e}")
            await interaction.response.send_message("❌ An error occurred while creating the reaction role!", ephemeral=True)
    
    @app_commands.command(name="removereactionrole", description="Remove a reaction role from a message")
    @app_commands.describe(
        channel="The channel where the message is located",
        message_id="The ID of the message",
        emoji="The emoji to remove"
    )
    @app_commands.default_permissions(manage_roles=True)
    async def remove_reaction_role(self, interaction: discord.Interaction, 
                                   channel: discord.TextChannel, message_id: str, emoji: str):
        """Remove a reaction role"""
        
        if not self.bot.db_pool:
            await interaction.response.send_message("❌ Database connection not available!", ephemeral=True)
            return
        
        try:
            async with self.bot.db_pool.acquire() as conn:
                # Find the reaction role
                reaction_role = await conn.fetchrow("""
                    SELECT * FROM role_reactions 
                    WHERE message_id = $1 AND emoji = $2 AND server_id = $3 AND is_active = true
                """, message_id, emoji, str(interaction.guild.id))
                
                if not reaction_role:
                    await interaction.response.send_message("❌ No reaction role found with that emoji on that message!", ephemeral=True)
                    return
                
                # Mark as inactive
                await conn.execute("""
                    UPDATE role_reactions 
                    SET is_active = false 
                    WHERE message_id = $1 AND emoji = $2
                """, message_id, emoji)
                
                # Try to remove the bot's reaction
                try:
                    message = await channel.fetch_message(int(message_id))
                    await message.clear_reaction(emoji)
                except:
                    pass  # If we can't remove it, that's okay
                
                embed = discord.Embed(
                    title="✅ Reaction Role Removed",
                    description=f"Removed {emoji} reaction role from message {message_id}",
                    color=discord.Color.green()
                )
                
                await interaction.response.send_message(embed=embed)
                print(f"Removed reaction role: {emoji} from message {message_id}")
                
        except Exception as e:
            print(f"Error removing reaction role: {e}")
            await interaction.response.send_message("❌ An error occurred while removing the reaction role!", ephemeral=True)
    
    @app_commands.command(name="listreactionroles", description="List all reaction roles in this server")
    async def list_reaction_roles(self, interaction: discord.Interaction):
        """List all active reaction roles"""
        
        if not self.bot.db_pool:
            await interaction.response.send_message("❌ Database connection not available!", ephemeral=True)
            return
        
        try:
            async with self.bot.db_pool.acquire() as conn:
                reaction_roles = await conn.fetch("""
                    SELECT * FROM role_reactions 
                    WHERE server_id = $1 AND is_active = true
                    ORDER BY created_at DESC
                """, str(interaction.guild.id))
                
                if not reaction_roles:
                    await interaction.response.send_message("❌ No reaction roles set up in this server!", ephemeral=True)
                    return
                
                embed = discord.Embed(
                    title="📋 Reaction Roles",
                    description=f"Active reaction roles in {interaction.guild.name}",
                    color=discord.Color.blue()
                )
                
                for rr in reaction_roles[:25]:  # Limit to 25 to avoid embed limits
                    role = interaction.guild.get_role(int(rr['role_id']))
                    role_name = role.mention if role else f"Deleted Role ({rr['role_id']})"
                    
                    channel = interaction.guild.get_channel(int(rr['channel_id']))
                    channel_name = channel.mention if channel else f"<#{rr['channel_id']}>"
                    
                    embed.add_field(
                        name=f"{rr['emoji']} → {role_name}",
                        value=f"Channel: {channel_name}\nMessage ID: {rr['message_id']}",
                        inline=False
                    )
                
                if len(reaction_roles) > 25:
                    embed.set_footer(text=f"Showing 25 of {len(reaction_roles)} reaction roles")
                
                await interaction.response.send_message(embed=embed)
                
        except Exception as e:
            print(f"Error listing reaction roles: {e}")
            await interaction.response.send_message("❌ An error occurred while fetching reaction roles!", ephemeral=True)
    
    @commands.Cog.listener()
    async def on_raw_reaction_add(self, payload: discord.RawReactionActionEvent):
        """Called when a reaction is added to a message"""
        
        # Ignore bot reactions
        if payload.user_id == self.bot.user.id:
            return
        
        if not self.bot.db_pool:
            return
        
        try:
            async with self.bot.db_pool.acquire() as conn:
                # Check if this is a reaction role
                reaction_role = await conn.fetchrow("""
                    SELECT * FROM role_reactions 
                    WHERE message_id = $1 AND emoji = $2 AND is_active = true
                """, str(payload.message_id), str(payload.emoji))
                
                if not reaction_role:
                    return
                
                # Get the guild and member
                guild = self.bot.get_guild(payload.guild_id)
                if not guild:
                    return
                
                member = guild.get_member(payload.user_id)
                if not member:
                    return
                
                # Get the role
                role = guild.get_role(int(reaction_role['role_id']))
                if not role:
                    print(f"Role {reaction_role['role_id']} not found")
                    return
                
                # Check if member already has the role
                if role in member.roles:
                    return
                
                # Add the role
                try:
                    await member.add_roles(role, reason="Reaction role")
                    print(f"Added role {role.name} to {member} via reaction")
                except discord.Forbidden:
                    print(f"Failed to add role {role.name} to {member}: Missing permissions")
                except Exception as e:
                    print(f"Error adding role: {e}")
                    
        except Exception as e:
            print(f"Error in on_raw_reaction_add: {e}")
    
    @commands.Cog.listener()
    async def on_raw_reaction_remove(self, payload: discord.RawReactionActionEvent):
        """Called when a reaction is removed from a message"""
        
        # Ignore bot reactions
        if payload.user_id == self.bot.user.id:
            return
        
        if not self.bot.db_pool:
            return
        
        try:
            async with self.bot.db_pool.acquire() as conn:
                # Check if this is a reaction role
                reaction_role = await conn.fetchrow("""
                    SELECT * FROM role_reactions 
                    WHERE message_id = $1 AND emoji = $2 AND is_active = true
                """, str(payload.message_id), str(payload.emoji))
                
                if not reaction_role:
                    return
                
                # Get the guild and member
                guild = self.bot.get_guild(payload.guild_id)
                if not guild:
                    return
                
                member = guild.get_member(payload.user_id)
                if not member:
                    return
                
                # Get the role
                role = guild.get_role(int(reaction_role['role_id']))
                if not role:
                    return
                
                # Check if member has the role
                if role not in member.roles:
                    return
                
                # Remove the role
                try:
                    await member.remove_roles(role, reason="Reaction role removed")
                    print(f"Removed role {role.name} from {member} via reaction removal")
                except discord.Forbidden:
                    print(f"Failed to remove role {role.name} from {member}: Missing permissions")
                except Exception as e:
                    print(f"Error removing role: {e}")
                    
        except Exception as e:
            print(f"Error in on_raw_reaction_remove: {e}")

async def setup(bot):
    await bot.add_cog(RoleReactionsCog(bot))

File: server/bot/commands/roleReactions.ts
import { 
  SlashCommandBuilder, 
  ChatInputCommandInteraction, 
  PermissionFlagsBits, 
  EmbedBuilder,
  ButtonBuilder,
  ButtonStyle,
  ActionRowBuilder,
  StringSelectMenuBuilder,
  StringSelectMenuOptionBuilder,
  ComponentType,
  TextChannel,
  MessageReaction,
  User,
  GuildMember,
  PartialMessageReaction,
  PartialUser
} from 'discord.js';
import { storage } from '../../storage';
import { error } from '../../utils/logger';

const roleReactionCommand = {
  data: new SlashCommandBuilder()
    .setName('rolereaction')
    .setDescription('Manage role reactions')
    .setDefaultMemberPermissions(PermissionFlagsBits.ManageRoles)
    .addSubcommand(subcommand =>
      subcommand
        .setName('create')
        .setDescription('Create a role reaction menu')
        .addStringOption(option =>
          option.setName('title')
            .setDescription('Title for the role menu')
            .setRequired(true)
            .setMaxLength(256))
        .addStringOption(option =>
          option.setName('description')
            .setDescription('Description for the role menu')
            .setRequired(false)
            .setMaxLength(2048))
        .addChannelOption(option =>
          option.setName('channel')
            .setDescription('Channel to post the role menu')
            .setRequired(false)))
    .addSubcommand(subcommand =>
      subcommand
        .setName('add')
        .setDescription('Add a role-emoji pair to an existing message')
        .addStringOption(option =>
          option.setName('message-id')
            .setDescription('Message ID to add reaction to')
            .setRequired(true))
        .addRoleOption(option =>
          option.setName('role')
            .setDescription('Role to assign')
            .setRequired(true))
        .addStringOption(option =>
          option.setName('emoji')
            .setDescription('Emoji to react with')
            .setRequired(true)))
    .addSubcommand(subcommand =>
      subcommand
        .setName('remove')
        .setDescription('Remove a role reaction')
        .addStringOption(option =>
          option.setName('message-id')
            .setDescription('Message ID')
            .setRequired(true))
        .addStringOption(option =>
          option.setName('emoji')
            .setDescription('Emoji to remove')
            .setRequired(true)))
    .addSubcommand(subcommand =>
      subcommand
        .setName('list')
        .setDescription('List all role reactions'))
    .addSubcommand(subcommand =>
      subcommand
        .setName('template')
        .setDescription('Create a role menu from preset templates')
        .addStringOption(option =>
          option.setName('type')
            .setDescription('Template type')
            .setRequired(true)
            .addChoices(
              { name: 'Gaming Roles', value: 'gaming' },
              { name: 'Notification Roles', value: 'notifications' },
              { name: 'Color Roles', value: 'colors' },
              { name: 'Hobby Roles', value: 'hobbies' }
            ))
        .addChannelOption(option =>
          option.setName('channel')
            .setDescription('Channel to post the template')
            .setRequired(false))),

  async execute(interaction: ChatInputCommandInteraction) {
    if (!interaction.guild) {
      await interaction.reply({ content: 'This command can only be used in a server.', ephemeral: true });
      return;
    }

    const subcommand = interaction.options.getSubcommand();

    try {
      switch (subcommand) {
        case 'create':
          await handleCreateRoleMenu(interaction);
          break;
        case 'add':
          await handleAddRoleReaction(interaction);
          break;
        case 'remove':
          await handleRemoveRoleReaction(interaction);
          break;
        case 'list':
          await handleListRoleReactions(interaction);
          break;
        case 'template':
          await handleCreateTemplate(interaction);
          break;
      }
    } catch (error) {
      error(`Error in rolereaction ${subcommand}:`, error);
      await interaction.reply({
        content: 'An error occurred while processing your request.',
        ephemeral: true
      });
    }
  }
};

async function handleCreateRoleMenu(interaction: ChatInputCommandInteraction) {
  const title = interaction.options.getString('title', true);
  const description = interaction.options.getString('description') || 'React to get roles!';
  const channel = interaction.options.getChannel('channel') as TextChannel || interaction.channel as TextChannel;

  const embed = new EmbedBuilder()
    .setColor(0x5865f2)
    .setTitle(title)
    .setDescription(description)
    .addFields({
      name: '📋 How to use',
      value: 'React with the emoji below to get the corresponding role. React again to remove the role.',
      inline: false
    })
    .setFooter({ text: 'Role reactions will be added by administrators' })
    .setTimestamp();

  try {
    const message = await channel.send({ embeds: [embed] });

    await interaction.reply({
      content: `✅ Role menu created! Message ID: \`${message.id}\`\nUse \`/rolereaction add\` to add role-emoji pairs.`,
      ephemeral: true
    });
  } catch (error) {
    await interaction.reply({
      content: 'Failed to create role menu. Check my permissions in the target channel.',
      ephemeral: true
    });
  }
}

async function handleAddRoleReaction(interaction: ChatInputCommandInteraction) {
  const messageId = interaction.options.getString('message-id', true);
  const role = interaction.options.getRole('role', true);
  const emoji = interaction.options.getString('emoji', true);

  if (!interaction.guild) return;

  try {
    // Find the message in the current channel first, then search other channels
    let message;
    try {
      message = await (interaction.channel as TextChannel).messages.fetch(messageId);
    } catch {
      // Search in other channels
      const channels = interaction.guild.channels.cache.filter(c => c.isTextBased());
      for (const [, channel] of channels) {
        try {
          message = await (channel as TextChannel).messages.fetch(messageId);
          break;
        } catch {
          continue;
        }
      }
    }

    if (!message) {
      await interaction.reply({ content: 'Message not found.', ephemeral: true });
      return;
    }

    // Check if bot can manage this role
    const botMember = await interaction.guild.members.fetch(interaction.client.user.id);
    if (role.position >= botMember.roles.highest.position) {
      await interaction.reply({ 
        content: 'I cannot assign this role because it is higher than or equal to my highest role.', 
        ephemeral: true 
      });
      return;
    }

    // Add reaction to message
    await message.react(emoji);

    // Save to database
    await storage.createRoleReaction({
      serverId: interaction.guild.id,
      channelId: message.channel.id,
      messageId: message.id,
      emoji: emoji,
      roleId: role.id,
    });

    // Update the embed to show the new role
    const embed = message.embeds[0];
    if (embed) {
      const updatedEmbed = EmbedBuilder.from(embed);
      
      // Add or update the roles field
      const existingFields = updatedEmbed.data.fields || [];
      let rolesField = existingFields.find(field => field.name === '🎭 Available Roles');
      
      if (!rolesField) {
        updatedEmbed.addFields({
          name: '🎭 Available Roles',
          value: `${emoji} - ${role.name}`,
          inline: false
        });
      } else {
        const currentRoles = rolesField.value;
        const newValue = `${currentRoles}\n${emoji} - ${role.name}`;
        
        // Update the field
        const fieldIndex = existingFields.findIndex(field => field.name === '🎭 Available Roles');
        existingFields[fieldIndex].value = newValue;
        updatedEmbed.setFields(existingFields);
      }

      await message.edit({ embeds: [updatedEmbed] });
    }

    await interaction.reply({
      content: `✅ Role reaction added! ${emoji} → ${role.name}`,
      ephemeral: true
    });

  } catch (error) {
    error('Error adding role reaction:', error);
    await interaction.reply({
      content: 'Failed to add role reaction. Make sure the emoji is valid and accessible.',
      ephemeral: true
    });
  }
}

async function handleRemoveRoleReaction(interaction: ChatInputCommandInteraction) {
  const messageId = interaction.options.getString('message-id', true);
  const emoji = interaction.options.getString('emoji', true);

  if (!interaction.guild) return;

  try {
    // Get role reaction from database
    const roleReaction = await storage.getRoleReaction(messageId, emoji);
    
    if (!roleReaction) {
      await interaction.reply({ content: 'Role reaction not found.', ephemeral: true });
      return;
    }

    // Find and update the message
    let message;
    try {
      const channel = interaction.guild.channels.cache.get(roleReaction.channelId) as TextChannel;
      message = await channel.messages.fetch(messageId);
    } catch {
      await interaction.reply({ content: 'Message not found.', ephemeral: true });
      return;
    }

    // Remove bot's reaction
    const reaction = message.reactions.cache.get(emoji);
    if (reaction) {
      await reaction.users.remove(interaction.client.user.id);
    }

    // Remove from database
    await storage.deleteRoleReaction(roleReaction.id);

    await interaction.reply({
      content: `✅ Role reaction removed for ${emoji}`,
      ephemeral: true
    });

  } catch (error) {
    error('Error removing role reaction:', error);
    await interaction.reply({
      content: 'Failed to remove role reaction.',
      ephemeral: true
    });
  }
}

async function handleListRoleReactions(interaction: ChatInputCommandInteraction) {
  if (!interaction.guild) return;

  try {
    // This would need a new storage method to get all role reactions for a server
    const embed = new EmbedBuilder()
      .setColor(0x5865f2)
      .setTitle('📋 Role Reactions')
      .setDescription('No role reactions found. Use `/rolereaction create` to get started!')
      .setTimestamp();

    await interaction.reply({ embeds: [embed], ephemeral: true });
  } catch (error) {
    await interaction.reply({
      content: 'Failed to list role reactions.',
      ephemeral: true
    });
  }
}

async function handleCreateTemplate(interaction: ChatInputCommandInteraction) {
  const templateType = interaction.options.getString('type', true);
  const channel = interaction.options.getChannel('channel') as TextChannel || interaction.channel as TextChannel;

  const templates = {
    gaming: {
      title: '🎮 Gaming Roles',
      description: 'React to get notified about your favorite games!',
      roles: [
        { emoji: '🎯', name: 'FPS Games' },
        { emoji: '⚔️', name: 'RPG Games' },
        { emoji: '🏎️', name: 'Racing Games' },
        { emoji: '🎲', name: 'Strategy Games' },
        { emoji: '🏹', name: 'Adventure Games' }
      ]
    },
    notifications: {
      title: '🔔 Notification Roles',
      description: 'Choose what notifications you want to receive!',
      roles: [
        { emoji: '📢', name: 'Announcements' },
        { emoji: '🎉', name: 'Events' },
        { emoji: '🎁', name: 'Giveaways' },
        { emoji: '📰', name: 'News' },
        { emoji: '🔔', name: 'General Updates' }
      ]
    },
    colors: {
      title: '🌈 Color Roles',
      description: 'Pick a color for your name!',
      roles: [
        { emoji: '🔴', name: 'Red' },
        { emoji: '🟠', name: 'Orange' },
        { emoji: '🟡', name: 'Yellow' },
        { emoji: '🟢', name: 'Green' },
        { emoji: '🔵', name: 'Blue' },
        { emoji: '🟣', name: 'Purple' },
        { emoji: '🩷', name: 'Pink' }
      ]
    },
    hobbies: {
      title: '🎨 Hobby Roles',
      description: 'Share your interests and hobbies!',
      roles: [
        { emoji: '🎨', name: 'Art & Design' },
        { emoji: '🎵', name: 'Music' },
        { emoji: '📚', name: 'Reading' },
        { emoji: '💻', name: 'Programming' },
        { emoji: '🏋️', name: 'Fitness' },
        { emoji: '🍳', name: 'Cooking' }
      ]
    }
  };

  const template = templates[templateType as keyof typeof templates];

  const embed = new EmbedBuilder()
    .setColor(0x5865f2)
    .setTitle(template.title)
    .setDescription(template.description)
    .addFields(
      {
        name: '📋 How to use',
        value: 'React with the emoji below to get the corresponding role. React again to remove the role.',
        inline: false
      },
      {
        name: '🎭 Available Roles',
        value: template.roles.map(role => `${role.emoji} - ${role.name}`).join('\n'),
        inline: false
      }
    )
    .setFooter({ text: 'Template created - roles need to be set up by administrators' })
    .setTimestamp();

  try {
    const message = await channel.send({ embeds: [embed] });

    // Add all reactions
    for (const role of template.roles) {
      await message.react(role.emoji);
    }

    await interaction.reply({
      content: `✅ ${template.title} template created! Message ID: \`${message.id}\`\n**Note:** You need to create the actual roles and link them using \`/rolereaction add\`.`,
      ephemeral: true
    });

  } catch (error) {
    error('Error creating template:', error);
    await interaction.reply({
      content: 'Failed to create template.',
      ephemeral: true
    });
  }
}

// Event handlers for role reactions
export async function handleReactionAdd(reaction: MessageReaction | PartialMessageReaction, user: User | PartialUser) {
  // Ignore bot reactions
  if (user.bot) return;

  try {
    // Fetch partial data if needed
    if (reaction.partial) {
      await reaction.fetch();
    }
    if (user.partial) {
      await user.fetch();
    }

    const guild = reaction.message.guild;
    if (!guild) return;

    // Get role reaction from database
    const roleReaction = await storage.getRoleReaction(reaction.message.id, reaction.emoji.name || reaction.emoji.toString());
    
    if (!roleReaction) return;

    // Get member and role
    const member = await guild.members.fetch(user.id);
    const role = guild.roles.cache.get(roleReaction.roleId);

    if (!role) {
      error(`Role ${roleReaction.roleId} not found for reaction ${roleReaction.emoji}`);
      return;
    }

    // Check if member already has the role
    if (member.roles.cache.has(role.id)) {
      return; // Already has role, no need to add again
    }

    // Add role
    await member.roles.add(role, 'Role reaction');

    // Try to send DM notification
    try {
      const dmEmbed = new EmbedBuilder()
        .setColor(0x00ff00)
        .setTitle('✅ Role Added')
        .setDescription(`You have been given the **${role.name}** role in **${guild.name}**!`)
        .setTimestamp();

      await user.send({ embeds: [dmEmbed] });
    } catch (error) {
      // User has DMs disabled, continue silently
    }

  } catch (error) {
    error('Error handling reaction add:', error);
  }
}

export async function handleReactionRemove(reaction: MessageReaction | PartialMessageReaction, user: User | PartialUser) {
  // Ignore bot reactions
  if (user.bot) return;

  try {
    // Fetch partial data if needed
    if (reaction.partial) {
      await reaction.fetch();
    }
    if (user.partial) {
      await user.fetch();
    }

    const guild = reaction.message.guild;
    if (!guild) return;

    // Get role reaction from database
    const roleReaction = await storage.getRoleReaction(reaction.message.id, reaction.emoji.name || reaction.emoji.toString());
    
    if (!roleReaction) return;

    // Get member and role
    const member = await guild.members.fetch(user.id);
    const role = guild.roles.cache.get(roleReaction.roleId);

    if (!role) {
      error(`Role ${roleReaction.roleId} not found for reaction ${roleReaction.emoji}`);
      return;
    }

    // Check if member has the role
    if (!member.roles.cache.has(role.id)) {
      return; // Doesn't have role, no need to remove
    }

    // Remove role
    await member.roles.remove(role, 'Role reaction removed');

    // Try to send DM notification
    try {
      const dmEmbed = new EmbedBuilder()
        .setColor(0xff6b6b)
        .setTitle('❌ Role Removed')
        .setDescription(`The **${role.name}** role has been removed from you in **${guild.name}**.`)
        .setTimestamp();

      await user.send({ embeds: [dmEmbed] });
    } catch (error) {
      // User has DMs disabled, continue silently
    }

  } catch (error) {
    error('Error handling reaction remove:', error);
  }
}

export const roleReactionCommands = [roleReactionCommand];